File: backend/src/main/java/com/dvmotos/service/ClienteService.java
package com.dvmotos.service;

import com.dvmotos.dto.request.ClienteRequest;
import com.dvmotos.dto.response.ClienteResponse;
import com.dvmotos.entity.Cliente;
import com.dvmotos.exception.BusinessException;
import com.dvmotos.exception.ResourceNotFoundException;
import com.dvmotos.repository.ClienteRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ClienteService {

    private final ClienteRepository clienteRepository;

    @Transactional(readOnly = true)
    public Page<ClienteResponse> findAll(String search, Pageable pageable) {
        Page<Cliente> clientes;
        if (search != null && !search.trim().isEmpty()) {
            clientes = clienteRepository.search(search.trim(), pageable);
        } else {
            clientes = clienteRepository.findByAtivoTrue(pageable);
        }

        return clientes.map(ClienteResponse::fromEntity);
    }

    public ClienteResponse findById(Long id) {
        Cliente cliente = clienteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente", id));
        return ClienteResponse.fromEntity(cliente);
    }

    @Transactional
    public ClienteResponse create(ClienteRequest request) {
        if (request.getCpfCnpj() != null && clienteRepository.existsByCpfCnpj(request.getCpfCnpj())) {
            throw new BusinessException("Já existe um cliente com este CPF/CNPJ");
        }

        Cliente cliente = Cliente.builder()
                .nome(request.getNome())
                .cpfCnpj(request.getCpfCnpj())
                .telefone(request.getTelefone())
                .email(request.getEmail())
                .endereco(request.getEndereco())
                .cidade(request.getCidade())
                .estado(request.getEstado())
                .cep(request.getCep())
                .observacoes(request.getObservacoes())
                .build();

        cliente = clienteRepository.save(cliente);
        return ClienteResponse.fromEntity(cliente);
    }

    @Transactional
    public ClienteResponse update(Long id, ClienteRequest request) {
        Cliente cliente = clienteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente", id));

        if (request.getCpfCnpj() != null && !request.getCpfCnpj().equals(cliente.getCpfCnpj())) {
            if (clienteRepository.existsByCpfCnpj(request.getCpfCnpj())) {
                throw new BusinessException("Já existe um cliente com este CPF/CNPJ");
            }
        }

        cliente.setNome(request.getNome());
        cliente.setCpfCnpj(request.getCpfCnpj());
        cliente.setTelefone(request.getTelefone());
        cliente.setEmail(request.getEmail());
        cliente.setEndereco(request.getEndereco());
        cliente.setCidade(request.getCidade());
        cliente.setEstado(request.getEstado());
        cliente.setCep(request.getCep());
        cliente.setObservacoes(request.getObservacoes());

        cliente = clienteRepository.save(cliente);
        return ClienteResponse.fromEntity(cliente);
    }

    @Transactional
    public void delete(Long id) {
        Cliente cliente = clienteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente", id));

        // Soft delete
        cliente.setAtivo(false);
        clienteRepository.save(cliente);
    }

    @Transactional
    public ClienteResponse activate(Long id) {
        Cliente cliente = clienteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente", id));
        cliente.setAtivo(true);
        cliente = clienteRepository.save(cliente);
        return ClienteResponse.fromEntity(cliente);
    }
}

File: frontend/src/app/core/services/cliente.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { environment } from "@env/environment";

export interface Cliente {
  id?: number;
  nome: string;
  cpfCnpj?: string;
  telefone?: string;
  email?: string;
  endereco?: string;
  cidade?: string;
  estado?: string;
  cep?: string;
  observacoes?: string;
  ativo?: boolean;
  createdAt?: string;
  totalVeiculos?: number;
}

export interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({
  providedIn: "root",
})
export class ClienteService {
  private apiUrl = `${environment.apiUrl}/clientes`;

  constructor(private http: HttpClient) {}

  findAll(
    page: number = 0,
    size: number = 20,
    search?: string,
  ): Observable<Page<Cliente>> {
    let params = new HttpParams()
      .set("page", page.toString())
      .set("size", size.toString());

    if (search) {
      params = params.set("search", search);
    }

    return this.http.get<Page<Cliente>>(this.apiUrl, { params });
  }

  findById(id: number): Observable<Cliente> {
    return this.http.get<Cliente>(`${this.apiUrl}/${id}`);
  }

  create(cliente: Cliente): Observable<Cliente> {
    return this.http.post<Cliente>(this.apiUrl, cliente);
  }

  update(id: number, cliente: Cliente): Observable<Cliente> {
    return this.http.put<Cliente>(`${this.apiUrl}/${id}`, cliente);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  activate(id: number): Observable<Cliente> {
    return this.http.patch<Cliente>(`${this.apiUrl}/${id}/activate`, {});
  }
}
